list.go

```diff
@@ -0,0 +1,92 @@
+package pack
+
+import (
+	"encoding/json"
+	"fmt"
+	"math/rand"
+	"reflect"
+
+	"github.com/renproject/surge"
+)
+
+type List struct {
+	T     Type
+	Elems []Value
+}
+
+func NewList(vs ...Value) (List, error) {
+	if len(vs) == 0 {
+		return List{}, nil
+	}
+
+	elems := make([]Value, len(vs))
+	var t Type
+	for i := range elems {
+		// Verify the list elements have a consistent type.
+		if t == nil {
+			t = vs[i].Type()
+		} else if vs[i].Type() != t {
+			return List{}, fmt.Errorf("inconsistent list type: expected %v, got %v", t, elems[i].Type())
+		}
+		elems[i] = vs[i]
+	}
+	return List{
+		T:     t,
+		Elems: elems,
+	}, nil
+}
+
+// Type returns the list type.
+func (v List) Type() Type {
+	return typeList{
+		Type: v.T,
+	}
```

Comment:
What happens when the list has length one? `T` is `nil`, so I am curious as to the behavior during encoding/decoding. If it is not an issue, why? And if it is, perhaps we should return an error for empty lists during `NewList` (or introduce some kind of `typeNil{}`).

Reply:
Why would `T` be `nil` if the list has one element?

Reply:
If you're referring to when it's an empty list, I've updated it to return an error.